bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Display.java

```diff
@@ -613,6 +615,63 @@ public void asyncExec (Runnable runnable) {
 	}
 }
 
+/**
+ * Executes the given runnable in the user-interface thread of this Display.
+ * <ul>
+ * <li>If the calling thread is the user-interface thread of this display it is
+ * executed immediately and the method returns after the command has run, any
+ * runtime exception thrown will be rethrown at the caller</li>
+ * <li>In all other cases the the <code>run()</code> method of the runnable is
+ * asynchronously executed as with the method
+ * {@link Display#asyncExec(Runnable)} at the next reasonable opportunity. The
+ * caller of this method continues to run in parallel, and is not notified when
+ * the runnable has completed.</li>
+ * </ul>
+ * <p>
+ * This can be used in cases where one want to execute some piece of code that
+ * should be guaranteed to run in the user-interface thread regardless of the
+ * current thread.
+ * </p>
+ *
+ * <p>
+ * Note that at the time the runnable is invoked, widgets that have the receiver
+ * as their display may have been disposed. Therefore, it is advised to check
+ * for this case inside the runnable before accessing the widget.
+ * </p>
+ * <p>
+ * Also note that most users would find it more convenient to use
+ * {@link Display#executor()} instead of acquiring a Display first and then use
+ * the Display as an Executor, for example when using
+ * <code>CompletableFuture</code>:
+ *
+ * <pre>
+ * CompletableFuture.supplyAsync(() -&gt; {
+ * 	// do something in the background
+ * }).thenAcceptAsync(computedValue -&gt; {
+ * 	// update the UI with the computed value
+ * }, Display.executor());
+ * </pre>
+ * </p>
+ *
+ * @param runnable the runnable to execute in the user-interface thread, never
+ *                 <code>null</code>
+ * @throws RejectedExecutionException if this task cannot be accepted for
+ *                                    execution
+ * @throws NullPointerException       if runnable is null
+ */
+@Override
+public void execute(Runnable runnable) {
+	Objects.requireNonNull(runnable);
+	if (isDisposed()) {
+		throw new RejectedExecutionException(new SWTException (SWT.ERROR_WIDGET_DISPOSED, null));
+	}
+	if (thread == Thread.currentThread()) {
+		runnable.run();
+	} else {
+		asyncExec(runnable);
+	}
```

Comment:
What i do not like at this implementation is the handling of unchecked Exceptions. Let there be a NPE in runnable.run() then it depends on the current thread if that NPE is thrown in the calling thread or not. That contradicts the usecase of "i dont care when it is run".

Reply:
That's consistent to how executors works and documented in the javadoc as well.

Reply:
I have now adopted what the AsyncHandler is doing by calling the error/runtime exception handlers of the display.

Reply:
the default handler will still just throw the exception, while asyncExec will catch it in org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(boolean)
```
       } catch (Throwable t) {
				SWT.error (SWT.ERROR_FAILED_EXEC, t);
	}
```
and then the SWT.error will wrap in SWTException .. and then the eventloop may care about the exception....

would it not better to simply always call asyncExec to get a consistent behavior?